PreMailer.Net/PreMailer.Net/Sources/LinkTagCssSource.cs

```
@@ -33,18 +39,112 @@ public string GetCss()
 
 			if (IsSupported(_downloadUri.Scheme))
 			{
-				try
+				return _cssContents ?? (_cssContents = DownloadContents());
+			}
+			return string.Empty;
+		}
+
+		private string DownloadContents()
```

Care to add some tests to safe-guard this logic? Doing that might also lead to a simplified design, perhaps it would be cleaner to keep the `import` logic in a class of its own and call it from here?

---

Do you perhaps have suggestions about the separate class? 
What I can think of is:
- It is not a ICssSource of it's own, so the class would not be something like ImportCssSource within the Sources namespace.
- It could be an extension within the Extensions namespace.
- Or it could be a helper class. But there are currently no other helper classes and namespace.

---

With tests I asume you mean in the PreMailer.Net.Tests project? I will have to find out how to use the mock or a mock to simulate downloading a css with imports.

---

I see that a separate class also makes testing cleaner. I first await your input on the separation step.

---

I extracted the import logic to a class ImportRuleCssSource. I now think it is a ICssSource of it's own. 
Because the class uses itself recursively it uses a StringBuilder instead of returning the contents from the GetCss method.

Still working on some tests. And do you have some suggestions on validation of the Css being added?

---

Content-type of the response is the most obvious. Otherwise, if the most common non-css imports are fonts - could it stop adding those?

Could also do some sanity checks of the contents, if it has `{` and `}` there's a high probability of it being CSS - although not safe.

---

> I extracted the import logic to a class ImportRuleCssSource

Cool, can you push those commits here then I can chip in also for more concrete input on tests.

---

I pushed the changes.
The current WebDownloader does not expose response headers. So, I cannot check the ContentType. I think the only best way of validating is just parsing like it is done now.

---

No, the WebDownloader would need changes. Also to make it more testable.

I'm not sure how the parser would behave here, if the root stylesheets imports two other CSS files where one of them is not CSS - would it fault all of it? It would be good to write unit tests for that.

---

I would leave the responsibility of failing to parse the complete css (including the imports) to the owner/author of the css files. In our case that's me :-). 

We have one main layout.css with some imports and a lot of css rules. Some of the import files also have import rules. Some referencing already imported files. Al those css files are parsed succesfully.
One of the import css files is a css with only variables. With the code from https://github.com/milkshakesoftware/PreMailer.Net/issues/339 in our wrapper class all variables are replaced with their corresponding values.

---

I created an ImportRuleCssSourceTest class and made some changes to the ImportRuleCssSource class. Do I have to commit and create another pull request? Or do I have to commit using the amend to your latest commit (Revert project file changes). I am not that experienced with github.

---

> I would leave the responsibility of failing to parse the complete css (including the imports) to the owner/author of the css files

The main concern is, this library is today out there and being used and is working in a certain way (a CSS with imports works fine, as the imports are ignored). If we introduce a change to this so that a CSS with imports working today will stop importing the entire CSS, it fails silently.

It should instead be opt in, so the default behavior is to let the root CSS be included but still ignore the failed imports - with the option to opt in for strict behavior.

---

> > I would leave the responsibility of failing to parse the complete css (including the imports) to the owner/author of the css files
> 
> The main concern is, this library is today out there and being used and is working in a certain way (a CSS with imports works fine, as the imports are ignored). If we introduce a change to this so that a CSS with imports working today will stop importing the entire CSS, it fails silently.
> 
> It should instead be opt in, so the default behavior is to let the root CSS be included but still ignore the failed imports - with the option to opt in for strict behavior.

I understand your concern and I would not like it either when something breaks because of a change.
Maybe the easiest way is to make processing import rules an opt in. By default import rules are ignored like it was before and with the opt in they are processed but with bad css potentially break the styling.

Or, I make the following changes:
- Change the ICssSource GetCss signature to something like GetCss(IList<string> cssSourceList) and all GetCss methods where the methods add the css to the list theirselfs. This way, the imported css can be added as separate blocks to the list so the block can be ignored when the parser fails because of bad css.
- Change PreMailer GetCssBlocks method in such a way it creates a list, walks the cssSources and calls the GetCss method with the list as parameter.
- Change the tests where GetCss method is called

---

It is better to encapsulate and do a proper validation. If you offer opt-in for imports, you expect it to work at least to the same degree as before.

If the imports results in their own CSS Block for the parser to process, that sounds fine. Instead of passing around lists, I would refactor `ICssSource` in such a way that it can return multiple strings e.g. as `IList<string>`.

---

I also have a suggestion to merge the CssSourceNodes and CssLinkNodes methods of PreMailer class.

If we use _document.QuerySelectorAll("style,link[rel=stylesheet]"), we get a collection of both stylesheet links and style nodes in the order they apear in the document. 

In the current implementation the css from a linked stylesheet overrules the css from the style nodes. IThe above approach is, I think, more compliant to the cascading of style(sheets). With this approach I can add a document style block after the general stylesheet in which I can overrule some general styles specific to emails.

If you wish I can add a change for this in this PR. Or would this change also be potentially breaking current PreMailer usage?

---

> If you wish I can add a change for this in this PR. Or would this change also be potentially breaking current PreMailer usage

It might, would rather address in separate PR and asses what it would do. Not entirely sure I follow without seeing some examples.

---

That's fine. I will wait till this PR is finished and then make a test case with example and propose the change in a new PR.

---

Thanks for the simplification and modifications.

---

> > If you wish I can add a change for this in this PR. Or would this change also be potentially breaking current PreMailer usage
> 
> It might, would rather address in separate PR and asses what it would do. Not entirely sure I follow without seeing some examples.

I already prepared a test method (ContainsLinkCssElement_DownloadsCss_InlinesContentCheckOrder) on the PreMailerTests class to show the issue with the current order in which the css of style and link nodes are handled.
I also prepared a new CssSourceNodesAll which selects style and link nodes together in the order they appear in the document.

Simply put, if currently a style block comes before or after a link node does not realy matter. The style block is selected first, thereafter the link node. This causes an incorrect preference to a stylesheet.

But this is for another PR. It's just that you know that when I have to commit something, you will see those new methods.